theano/tensor/nnet/abstract_conv.py

```diff
@@ -718,10 +751,17 @@ def conv2d(self, img, kern, mode="valid"):
             for b in xrange(img.shape[0]):
                 for n in xrange(kern.shape[0]):
                     for im0 in xrange(img.shape[1]):
-                        # some cast generates a warning here
-                        out[b, n, ...] += _convolve2d(img[b, im0, ...],
-                                                      kern[n, im0, ...],
-                                                      1, val, bval, 0)
+                        for row_offset in xrange(dilation[0]):
+                            for col_offset in xrange(dilation[1]):
+                                fragment = img[b, im0,
+                                               row_offset::dilation[0],
+                                               col_offset::dilation[1]]
```

Comment:
Yeah well didn't consider speed given it's a debug mode implementation. I've changed the implementation and will push it today. As for the optimizer, I was **wrong**  , it does use the gemm version correctly. It seems that in my script gemm was excluded (without noticing it). So i created a new clean script and indeed it used the gemm version when dilation != (1, 1). Will post the updates later today.